src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -488,8 +492,9 @@ public async Task OnCompilationEventProcessedAsync(CompilationEvent compilationE
         /// <summary>
         /// Gets pending events for given set of analyzers for the given syntax tree.
         /// </summary>
-        public ImmutableArray<CompilationEvent> GetPendingEvents(ImmutableArray<DiagnosticAnalyzer> analyzers, SyntaxTree tree)
+        public ImmutableArray<CompilationEvent> GetPendingEvents(ImmutableArray<DiagnosticAnalyzer> analyzers, SyntaxTree tree, CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
same goes for all these other cancellation check.

another way, you can do is using semaphoreSlim rather than using lock since semaphoreSim you can give cancellation token to the contstruct. then, when cancellation happened, the lock itself will throw and unlock itself, rather than waiting the lock to released and then check cancellation and then throw.

…

SemaphoreSlim internally just spin wait to check cancellation time to time so, checking cancellation after _gate is probably all we need?